File: AnalisadorLexico.java
import java.io.FileReader;
import java.io.IOException;

public class AnalisadorLexico implements Constantes {

    protected String nome;
    protected char proximoCaractere;
    protected int linha = 1;
    protected StringBuffer entrada = new StringBuffer();
    protected int posicao = 0;
    protected Token tokenReconhecido;

    public AnalisadorLexico(String nome) {
        this.nome = nome;
        try {
            FileReader file = new FileReader(nome);
            int c;
            while((c = file.read()) != -1) {
                this.entrada.append((char) c);
            }
            file.close();
            leProxCaractere();
        }

        catch (IOException erro) {
            throw new RuntimeException("Erro de leitura no arquivo" + nome);
        }
    }

    public AnalisadorLexico() {
        this.nome = "entrada.txt";
        try {
            FileReader file = new FileReader(nome);
            int c;
            while((c = file.read()) != -1) {
                this.entrada.append((char) c);
            }
            file.close();
            leProxCaractere();
        }

        catch (IOException erro) {
            throw new RuntimeException("Erro de leitura no arquivo" + nome);
        }
    }

    public void leProxCaractere() {
        try {
            this.proximoCaractere = this.entrada.charAt(this.posicao);
            posicao++;
        }

        catch (IndexOutOfBoundsException e) {
            this.proximoCaractere = EOF.charAt(0);
        }
    }

    public boolean proximoCaractereIs(String s) {
        return (s.indexOf(this.proximoCaractere) != -1);
    }
}

File: analisadorLexicoC++/AnalisadorLexico.cpp
#include "AnalisadorLexico.h"

#include <fstream>
#include <stdexcept>

AnalisadorLexico::AnalisadorLexico(const std::string& nome)
    : nome(nome),
      proximoCaractere(Constantes::EOF_CHAR),
      linha(1),
      entrada(""),
      posicao(0)
{
    std::ifstream arquivo(nome);

    if (!arquivo.is_open()) {
        throw std::runtime_error(
            "Erro de leitura no arquivo " + nome
        );
    }

    char c;

    while (arquivo.get(c)) {
        entrada += c;
    }

    arquivo.close();

    leProxCaractere();
}

AnalisadorLexico::AnalisadorLexico()
    : AnalisadorLexico("entrada.txt")
{
}

void AnalisadorLexico::leProxCaractere() {
    if (posicao < static_cast<int>(entrada.size())) {

        proximoCaractere = entrada[posicao];

        if (proximoCaractere == '\n') {
            linha++;
        }

        posicao++;
    }
    else {
        proximoCaractere = Constantes::EOF_CHAR;
    }
}

bool AnalisadorLexico::proximoCaractereIs(
        const std::string& s) const {

    return s.find(proximoCaractere) != std::string::npos;
}

Constantes::Token
AnalisadorLexico::getTokenReconhecido() const {
    return tokenReconhecido;
}

std::string AnalisadorLexico::getNome() const {
    return nome;
}